File: algorithms/breadthfirstsearch.cpp
//////////////////////////////////////////////////////////////////////
// Author: Arthur Anthony da Cunha Romão E Silva                   ///
// Enrollment: 20190011208                                         ///
//                                                                 ///
//////////////////////////////////////////////////////////////////////


#include <iostream>
#include <queue>

using namespace std;

void addEdge(int** mat, int u, int v){
	mat[u - 1][v - 1] = 1;
	mat[v - 1][u - 1] = 1;
}

void show(int **mat, int length){
	for(int i=0;i<length;i++){
		for(int j=0;j<length;j++){
			cout << mat[i][j] << " ";
		}
		cout << endl;
	}
}

void bfs(int** mat, int length, int* visited, int u){
	cout << u << " ";
	queue<int> Q;
	Q.push(u - 1);
	visited[u -1] = 1;
	
	while(Q.size() > 0){
		u = Q.front();
		for(int v=0;v<length;v++){
			if(mat[u][v] == 1 && visited[v] == 0){
				visited[v] = 1;
				Q.push(v);
				cout << v + 1 << " ";
			}
		}
		Q.pop();
	}
	cout << endl;
}

int main(int argc, char* argv[]){
	int length = 7;
	int** mat = (int**) calloc(length, sizeof(int*));
	int *visited = (int*) calloc(length, sizeof(int));

	for(int i=0;i<length;i++)
		mat[i] = (int*) calloc(length, sizeof(int));	

	addEdge(mat, 1, 2);
	addEdge(mat, 2, 3);
	addEdge(mat, 3, 4);
	addEdge(mat, 4, 5);
	addEdge(mat, 5, 6);
	addEdge(mat, 6, 7);


	bfs(mat, length, visited, 1);
	cout << endl;
	return 0;
}

File: algorithms/breadthfirstsearch.py
######################################################################
## Author: Arthur Anthony da Cunha Romão E Silva                   ###
## Enrollment: 20190011208                                         ###
##                                                                 ###
######################################################################

class Graph(object):
	def __init__(self, length):
		self.length = length
		self.vertices = [[0] * length for x in range(length)]
		self.visited = [0] * length

	def addEdge(self, u, v):
		self.vertices[u - 1][v - 1] = 1
		self.vertices[v - 1][u - 1] = 1

	def show(self):
		for x in self.vertices:
			for y in x:
				print(y, end=" ")
			print("")

	def bfs(self, u):
		print(u, end=" ")
		queue = [u - 1]
		self.visited[u - 1] = 1

		while len(queue) > 0:
			u = queue[0]
			for v in range(self.length):
				if(self.vertices[u][v] == 1 and self.visited[v] == 0):
					self.visited[v] = 1
					queue.append(v)					
					print(v + 1, end=" ")
			queue.pop(0)
		print("")



g = Graph(7)
g.addEdge(1, 2)
g.addEdge(2, 3)
g.addEdge(3, 4)
g.addEdge(4, 5)
g.addEdge(5, 6)
g.addEdge(6, 7)
g.bfs(1)

print()